### PR TITLE
fix py_cast

### PR DESCRIPTION
**Pull request by songjhaha**
fix https://github.com/Suzhou-Tongyuan/jnumpy/issues/29

### CODE REVIEW

**Comment by songjhaha** [comment]
fix https://github.com/Suzhou-Tongyuan/jnumpy/issues/32 by https://github.com/Suzhou-Tongyuan/jnumpy/pull/31/commits/3415b637325d5c21c4a1eca543dcdef23c777daa

File: TyPython/src/CPython.ORM.jl [hunk 1]
```diff
@@ -210,7 +210,7 @@ end
     result = Expr(:tuple, [:($py_coerce($t, $unsafe_pytuple_get(py, $(i - 1)))) for (i, t) in enumerate(T.parameters)]...)
     quote
         if $CPython.PyAPI.PyObject_IsInstance(py,  $CPython.PyAPI.PyTuple_Type) == 0
-            $CPython.PyAPI.PyErr_SetString($CPython.PyAPI.PyExc_TypeError[], "expected a tuple")
+            $py_seterror!(G_PyBuiltin.TypeError, "expected a tuple")
```
Comment (thautwarm) [inline TyPython/src/CPython.ORM.jl:213]: ~~We shall not use `G_PyBuiltin` here.~~

~~Probably `$get_py_builtin().TypeError`.~~

File: TyPython/src/CPython.ORM.jl [hunk 2]
```diff
@@ -267,8 +267,16 @@ function py_cast(::Type{T}, py::Py)::T where T <: AbstractString
     py_coerce(T, py)
 end
 
-function py_cast(::Type{T}, py::Py)::T where T <: Tuple
-    py_coerce(T, py)
+@generated function py_cast(::Type{T}, py::Py)::T where T <: Tuple
+    T isa DataType && :(error("$T is not a data type"))
+    result = Expr(:tuple, [:($py_cast($t, $unsafe_pytuple_get(py, $(i - 1)))) for (i, t) in enumerate(T.parameters)]...)
+    quote
+        if $CPython.PyAPI.PyObject_IsInstance(py,  $CPython.PyAPI.PyTuple_Type) == 0
```
Comment (thautwarm) [inline TyPython/src/CPython.ORM.jl:274]: Checking the length is required. The same fix should be applied to `py_coerce`.

File: TyPython/src/CPython.ORM.jl (hunk 1 above)
Comment (thautwarm) [reply TyPython/src/CPython.ORM.jl:213]: P.S: It's okay to use `G_PyBuiltin` here as it is inside the AST returned by the generated function. 

This means the scope to resolve symbols is always the module that the generated function is defined in.

**Review by thautwarm — CHANGES_REQUESTED** [review]

**Review by thautwarm — APPROVED** [review]

**Comment by thautwarm** [comment]
You can now merge this PR for free.

**Pull request merged**